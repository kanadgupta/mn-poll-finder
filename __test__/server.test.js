const app = require('../server');

describe('fastify server tests', () => {
  describe('GET /', () => {
    it('should return 200 with standard template', async () => {
      const response = await app.inject({
        method: 'GET',
        url: '/',
      });

      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(200);
    });

    it('should return 200 when requesting JSON via Accept header', async () => {
      const response = await app.inject({
        method: 'GET',
        url: '/',
        headers: { Accept: 'application/json' },
      });

      expect(response.body).toBe('{}');
      expect(response.statusCode).toBe(200);
    });

    it('should return 200 when requesting JSON via query param', async () => {
      const response = await app.inject({
        method: 'GET',
        url: '/?format=json',
        headers: { Accept: 'application/json' },
      });

      expect(response.body).toBe('{}');
      expect(response.statusCode).toBe(200);
    });
  });

  describe('GET /?example=go', () => {
    it('should return 200 with standard template', async () => {
      const response = await app.inject({
        method: 'GET',
        url: '/?example=go',
        headers: { Accept: 'application/json' },
      });

      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(200);
    });

    it('should return 200 when requesting JSON', async () => {
      const response = await app.inject({
        method: 'GET',
        url: '/?example=go',
        headers: { Accept: 'application/json' },
      });

      expect(response.body).toBe(
        JSON.stringify({
          address: '2506 Central Ave NE, Minneapolis, MN 55418',
          gmaps: 'https://www.google.com/maps/search/?api=1&query=2506+Central+Ave+NE%2C+Minneapolis%2C+MN+55418',
          precinct: {
            Precinct: 'Minneapolis W-1 P-9',
            PrecinctID: '270531400',
            County: 'Hennepin',
            CountyID: '27',
            CongDist: '5',
            MNSenDist: '60',
            MNLegDist: '60A',
            CtyComDist: '2',
            Hospital: 'no data',
            Judicial: '04',
            MCDCode: '135',
            MCDName: 'Minneapolis',
            Park: '1',
            PrecinctCode: '1400',
            SoilAndWater: 'no data',
            Ward: 'W-01',
          },
          mplsPollingPlace21: {
            address: '2030 Monroe St NE',
            building: 'Edison High School (Gym lobby)',
            directions: '',
            gmapsUrl:
              'https://www.google.com/maps/search/?api=1&query=Edison+High+School+%28Gym+lobby%29+2030+Monroe+St+NE&query_place_id=ChIJx58zzJIts1IRDTGIN8Fm7tk',
          },
        })
      );
      expect(response.statusCode).toBe(200);
    });
  });
});
