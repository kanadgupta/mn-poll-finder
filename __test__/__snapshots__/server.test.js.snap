// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fastify server tests GET / should return 200 with standard template 1`] = `
"
<html lang='en'>
  <head>
    <meta charset='utf-8' />
    <link rel='icon' href='https://fav.farm/üó≥Ô∏è' />
    <meta name='viewport' content='width=device-width, initial-scale=1' />

    <title>MPLS 2021 Poll Finder</title>

    <!-- Meta tags for SEO and social sharing -->
    <link rel='canonical' href='https://mpls.vote' />
    <meta name='description' content='An open-source, API-backed poll finder that (mostly) doesn&#x27;t suck ‚ù§Ô∏è' />
    <meta property='og:title' content='MPLS 2021 Poll Finder' />
    <meta property='og:image' content='https://emojipedia-us.s3.dualstack.us-west-1.amazonaws.com/thumbs/240/apple/285/ballot-box-with-ballot_1f5f3-fe0f.png' />
    <meta property='og:type' content='article' />
    <meta property='og:url' content='https://mpls.vote' />
    <meta property='og:description' content='An open-source, API-backed poll finder that (mostly) doesn&#x27;t suck ‚ù§Ô∏è' />
    <meta name='twitter:card' content='summary' />

    <!-- Import the webpage's stylesheet -->
    <link rel='stylesheet' href='/style.css' />
  </head>
  <body>
    <div class='wrapper'>
      <div class='content' role='main'>

        <!-- This is the start of content for our page -->
        <h1 class='title'><a class='title' href='/'>MPLS 2021 Poll Finder</a></h1>

        <!-- Add text indicating that we've found the precinct OR had an error -->


            <p class='intro subtitle'>
              An open-source, API-backed poll finder that (mostly) doesn't suck ‚ù§Ô∏è
            </p><p class='intro'>
              ‚ö†Ô∏è
              <a href='#faq'>Check out the FAQ</a>
              before diving in!
            </p>
          <p class='intro'>üëÄ Need an example?
            <a href='?example=go'>Click here!</a></p>

        <div class='form-wrapper'>

          <!-- Our default paragraph/message -->
          <form class='address-form' method='get'>
            <label for='address-input'>
                  <span class='subtitle'>What's your home address? üè°</span><br />
              <input
                id='address-input'
                name='address'
                required='required'
                placeholder='123 Main St, 55411'
                type='text'
                value=''
              />
            </label>
            <button type='submit'>Search</button>
          </form>
        </div>

        <!-- FAQ -->
        <div class='faq-wrapper'>
          <h2 id='faq'>
            Frequently Asked Questions (F.A.Q.)
          </h2>
          <h3>What is this?</h3>
          <p>
            This is a tool that (mostly) helps you with one thing: finding your polling place (if you live in
            Minneapolis) and your voting precinct (if you live in Minnesota)!
          </p>
          <h3>Are there any limitations with this tool?</h3>
          <p>Yes. lol.
          </p>
          <p>First off, this relies on Google Maps for geocoding the data. Like any service, if you send garbage data to
            it, you'll receive garbage in return. In other words, you'll see the best results if you send complete
            street addresses (that actually exist!), with brownie points if you include a ZIP code.</p>
          <p>With any geocoder, the data tends to be less accurate in less populous areas. While it should work just
            fine for most parts of the state (provided you send it complete and valid addresses), take it with a grain
            of salt if you're outside of the Twin Cities Metro Area!
            <b>If you're ever unsure of the results you're getting,
              <a href='https://pollfinder.sos.state.mn.us/' rel='noopener' target='_blank'>the official poll finder</a>
              is your best bet.</b></p>
          <p><i>And lastly... this tool doesn't have any sample ballot data but long story short: #DontRankFrey and vote
              #NoYesYes ‚ù§Ô∏è</i></p>
          <h3>So this is cool and all but... the Minnesota Secretary of State
            <a href='https://pollfinder.sos.state.mn.us/' rel='noopener' target='_blank'>already does this</a>. Why are
            you?</h3>
          <p>It's true‚Äîthe Secretary of State (SOS) has done excellent work on these fronts and it serves us very well!
            However, the current system requires
            <b>a lot</b>
            of clicks to access all of that valuable information‚Äînot just access to your precinct and its corresponding
            polling place, but all of your electoral boundaries (Ward, County, Congressional Districts, etc.) and
            everything on your ballot.
          </p>
          <p>And what's worse, there's no discernable pattern to these URLs, meaning that all of this information is
            difficult to access at scale without significant resources. This makes it harder for grassroots campaigns to
            do their jobs, which in turn affects electoral turnout.</p>
          <p>This tool isn't intended to replace the SOS's system‚Äîit was built by a single person, I honestly couldn't
            do that if I tried! The purpose of this site is to demonstrate how modern web technologies can be used to
            create API-driven tools, which can (hopefully!) reduce some of the barriers to providing information for
            people. üå±
          </p>
          <h3>You mentioned that it's API-driven! How does that work?</h3>
          <p>You can retrieve a JSON version of data provided on the website by passing an
            <code>Accept</code>
            header in your request that specifies
            <code>application/json</code>. You can also pass a
            <code>format=json</code>
            query parameter (<a href='/?example=go&format=json' rel='noopener' target='_blank'>like this!</a>) and that
            should also do the trick.
          </p>
          <p>The markup on this webpage also includes a few hidden tricks for Google Sheets users to easily integrate
            this data into their workflows. You can grab the XPath selectors of certain fields on the page and then use
            formulas like these to import the data:</p>
          <ul>
            <li><code>=IMPORTXML(\\"https://mpls.vote?address=\\"&ENCODEURL(\\"100 Main St\\"),\\"//*[@id='Precinct']\\")</code></li>
            <li><code>=IMPORTXML(\\"https://mpls.vote?address=\\"&ENCODEURL(\\"100 Main St\\"),\\"//*[@id='address']\\")</code></li>
            <li><code>=IMPORTXML(\\"https://mpls.vote?address=\\"&ENCODEURL(\\"100 Main St\\"),\\"//*[@id='mplsPollingPlace21']\\")</code></li>
          </ul>
          <p>API docs TBA! ü¶â</p>
          <h3>Do you have a privacy policy?</h3>
          <p>This is a humble proof-of-concept so it doesn't have a formal privacy policy, but this application does not
            load with any invasive analytics/tracking scripts, nor does it log any of the information you send. Oh, and
            the code is
            <a href='https://github.com/kanadgupta/mn-precinct-finder' rel='noopener' target='_blank'>fully open-source</a>
            if you'd like to take a look!</p>
          <p>The information you submit to the site via the query parameters (address information, geographic
            coordinates via the 'Current Location' button, etc.) is anonymously sent to the
            <a href='https://developers.google.com/maps/documentation/geocoding' rel='noopener' target='_blank'>Google
              Maps Geocoding API</a>
            so that information is subject to
            <a href='https://policies.google.com/privacy' rel='noopener' target='_blank'>their privacy policy</a>
            and
            <a href='https://policies.google.com/terms' rel='noopener' target='_blank'>their terms of use</a>. So don't,
            like, send your social security number through the form? This site won't see it or log it anywhere, but it's
            technically being sent to Google.
          </p>
          <h3>Will elections save us?</h3>
          <p>lol... you must be new here</p>
        </div>
      </div>
      <footer class='footer'><span>made with &hearts; by
          <a href='https://twitter.com/kanadgupta' rel='noopener' target='_blank'>kanad</a></span></footer>
    </div>
    <script>
      function toggleInfo(e) {
        e.preventDefault();
        // Toggle display of every hidden item
        document.querySelectorAll('li.extra').forEach(item => {
          item.classList.toggle('hidden');
        });
        // Update attributes on toggle element
        const isHidden = document.querySelector('li.extra').classList.contains('hidden');
        e.target.innerText = isHidden ? 'Show irrelevant info...' : 'Hide irrelevant info...';
        e.target.setAttribute('aria-expanded', !isHidden);
      }
    </script>
  </body>
</html>
"
`;

exports[`fastify server tests GET /?example=go should return 200 with standard template 1`] = `"{\\"address\\":\\"2506 Central Ave NE, Minneapolis, MN 55418\\",\\"gmaps\\":\\"https://www.google.com/maps/search/?api=1&query=2506+Central+Ave+NE%2C+Minneapolis%2C+MN+55418\\",\\"precinct\\":{\\"Precinct\\":\\"Minneapolis W-1 P-9\\",\\"PrecinctID\\":\\"270531400\\",\\"County\\":\\"Hennepin\\",\\"CountyID\\":\\"27\\",\\"CongDist\\":\\"5\\",\\"MNSenDist\\":\\"60\\",\\"MNLegDist\\":\\"60A\\",\\"CtyComDist\\":\\"2\\",\\"Hospital\\":\\"no data\\",\\"Judicial\\":\\"04\\",\\"MCDCode\\":\\"135\\",\\"MCDName\\":\\"Minneapolis\\",\\"Park\\":\\"1\\",\\"PrecinctCode\\":\\"1400\\",\\"SoilAndWater\\":\\"no data\\",\\"Ward\\":\\"W-01\\"},\\"mplsPollingPlace21\\":{\\"address\\":\\"2030 Monroe St NE\\",\\"building\\":\\"Edison High School (Gym lobby)\\",\\"directions\\":\\"\\",\\"gmapsUrl\\":\\"https://www.google.com/maps/search/?api=1&query=Edison+High+School+%28Gym+lobby%29+2030+Monroe+St+NE&query_place_id=ChIJx58zzJIts1IRDTGIN8Fm7tk\\"}}"`;
